core/src/state/world_state.rs

```diff
@@ -47,25 +56,51 @@ impl StateHandle {
     }
 
     pub fn state(&self) -> RwLockReadGuard<WorldState> {
-        self.0.read().expect("Could not acquire world_state lock.")
+        self.state
+            .read()
+            .expect("Could not acquire world_state lock.")
     }
 
     pub(crate) fn write_state(&self) -> RwLockWriteGuard<WorldState> {
-        self.0.write().expect("Could not acquire world_state lock.")
+        self.state
+            .write()
+            .expect("Could not acquire world_state lock.")
+    }
+
+    /** Block asynchronously until the given sequence number is reached. */
+    pub async fn wait_for_seq(&self, sequence: u64) {
+        if self.state().logical_time >= sequence {
+            return;
+        }
+        let notify = Arc::new(Notify::new());
+        self.write_state().add_listener(sequence, notify.clone());
+        notify.notified().await;
     }
 }
 
 #[derive(Default, Debug)]
 pub struct WorldState {
     logical_time: u64,
     pub clusters: BTreeMap<ClusterName, ClusterState>,
+    listeners: Arc<DashMap<u64, Vec<Arc<Notify>>>>,
```

Comment:
Does this need to be in an arc?

Reply:
Also if you use notify_all should be able to get rid of the Vec.

Reply:
I couldn't figure out what you mean by notify_all but I realized I could use a broadcast channel so I did that.

Good catch, it doesn't need to be an arc, or a dashmap for that matter.

Reply:
https://docs.rs/tokio/latest/tokio/sync/struct.Notify.html#method.notify_waiters

my b got the method name wrong, but a broadcast channel is fine too.

Reply:
ooh yeah I like that, it feels lighter-weight

Reply:
Idk about weight, I'd bet they're basically the same efficiency wise, easier to read with the notify though. your call.